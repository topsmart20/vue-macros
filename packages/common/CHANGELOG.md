# @vue-macros/common

## 0.12.2

### Patch Changes

- [#108](https://github.com/sxzz/unplugin-vue-macros/pull/108) [`d036273`](https://github.com/sxzz/unplugin-vue-macros/commit/d036273e88368d6b6ef33a31b7bc13f2c0c0f005) Thanks [@renovate](https://github.com/apps/renovate)! - upgrade deps

## 0.12.1

### Patch Changes

- [#104](https://github.com/sxzz/unplugin-vue-macros/pull/104) [`97d82c6`](https://github.com/sxzz/unplugin-vue-macros/commit/97d82c62a3aad1b1158b9697ad6bd9d052c1c086) Thanks [@sxzz](https://github.com/sxzz)! - add runtime defineModel, requires `@vueuse/core`.

  ⚠️ BREAKING CHANGE: original `defineModel` renamed to `$defineModel`.

## 0.12.0

### Minor Changes

- [`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72) Thanks [@sxzz](https://github.com/sxzz)! - sync version

### Patch Changes

- [`f3a1b08`](https://github.com/sxzz/unplugin-vue-macros/commit/f3a1b08f49060740a5e53a8f3e91db0dc47c366f) Thanks [@sxzz](https://github.com/sxzz)! - extract walk ast to common

## 0.12.0-beta.3

### Patch Changes

- [`f3a1b08`](https://github.com/sxzz/unplugin-vue-macros/commit/f3a1b08f49060740a5e53a8f3e91db0dc47c366f) Thanks [@sxzz](https://github.com/sxzz)! - extract walk ast to common

## 0.12.0-beta.2

### Minor Changes

- [`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72) Thanks [@sxzz](https://github.com/sxzz)! - sync version
