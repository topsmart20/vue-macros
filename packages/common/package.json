{
  "name": "@vue-macros/common",
  "version": "0.9.1",
  "packageManager": "pnpm@7.9.0",
  "license": "MIT",
  "homepage": "https://github.com/sxzz/unplugin-vue-macros#readme",
  "bugs": {
    "url": "https://github.com/sxzz/unplugin-vue-macros/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sxzz/unplugin-vue-macros.git"
  },
  "author": "三咲智子 <sxzz@sxzz.moe>",
  "files": [
    "dist"
  ],
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsup && tsx ../../scripts/postbuild.mts",
    "dev": "tsup --watch"
  },
  "peerDependencies": {
    "vue": "^2.7.0 || ^3.2.25"
  },
  "dependencies": {
    "@babel/types": "^7.18.10",
    "@vue/compiler-sfc": "^3.2.37",
    "magic-string": "^0.26.2"
  },
  "devDependencies": {
    "@babel/parser": "^7.18.11",
    "tsup": "^6.2.2",
    "vue": "^3.2.37"
  },
  "engines": {
    "node": ">=14.19.0"
  }
}
