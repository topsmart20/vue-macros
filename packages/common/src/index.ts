export * from './vue'
export * from './ast'
export * from './constants'
export * from './unplugin'
export { default as MagicString } from 'magic-string'
