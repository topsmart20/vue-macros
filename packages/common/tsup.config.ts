import { readFile } from 'node:fs/promises'
import { defineConfig } from 'tsup'

const rawRE = /[&?]raw(?:&|$)/

export default defineConfig({
  entry: ['./src/*.ts'],
  format: ['cjs', 'esm'],
  target: 'node14',
  splitting: true,
  watch: !!process.env.DEV,
  dts: process.env.DEV
    ? false
    : {
        compilerOptions: {
          paths: {},
          composite: false,
        },
      },
  tsconfig: '../../tsconfig.lib.json',
  clean: true,
  esbuildPlugins: [
    {
      name: 'raw-plugin',
      setup(build) {
        build.onLoad({ filter: /.*/ }, async ({ path, suffix }) => {
          if (!rawRE.test(suffix)) return
          // raw query, read file and return as string
          return {
            contents: `export default ${JSON.stringify(
              await readFile(path, 'utf-8')
            )}`,
          }
        })
      },
    },
  ],
})
