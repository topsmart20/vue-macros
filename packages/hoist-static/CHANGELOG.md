# @vue-macros/hoist-static

## 0.12.3

### Patch Changes

- Updated dependencies [[`6fa5b3e`](https://github.com/sxzz/unplugin-vue-macros/commit/6fa5b3eed9de69a19921f0b38809d80d2cde3d50), [`1ac6e49`](https://github.com/sxzz/unplugin-vue-macros/commit/1ac6e496b3039a78f0288ca53698f88461ab6efc)]:
  - @vue-macros/common@0.12.3

## 0.12.2

### Patch Changes

- [#108](https://github.com/sxzz/unplugin-vue-macros/pull/108) [`d036273`](https://github.com/sxzz/unplugin-vue-macros/commit/d036273e88368d6b6ef33a31b7bc13f2c0c0f005) Thanks [@renovate](https://github.com/apps/renovate)! - upgrade deps

- [`0d46d80`](https://github.com/sxzz/unplugin-vue-macros/commit/0d46d801aea08a73c8e637d2a2ffabbd03565b37) Thanks [@sxzz](https://github.com/sxzz)! - keep empty <script setup>

- Updated dependencies [[`d036273`](https://github.com/sxzz/unplugin-vue-macros/commit/d036273e88368d6b6ef33a31b7bc13f2c0c0f005)]:
  - @vue-macros/common@0.12.2

## 0.12.1

### Patch Changes

- Updated dependencies [[`97d82c6`](https://github.com/sxzz/unplugin-vue-macros/commit/97d82c62a3aad1b1158b9697ad6bd9d052c1c086)]:
  - @vue-macros/common@0.12.1

## 0.12.0

### Minor Changes

- [`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72) Thanks [@sxzz](https://github.com/sxzz)! - sync version

### Patch Changes

- Updated dependencies [[`f3a1b08`](https://github.com/sxzz/unplugin-vue-macros/commit/f3a1b08f49060740a5e53a8f3e91db0dc47c366f), [`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72)]:
  - @vue-macros/common@0.12.0

## 0.12.0-beta.3

### Patch Changes

- Updated dependencies [[`f3a1b08`](https://github.com/sxzz/unplugin-vue-macros/commit/f3a1b08f49060740a5e53a8f3e91db0dc47c366f)]:
  - @vue-macros/common@0.12.0-beta.3

## 0.12.0-beta.2

### Minor Changes

- [`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72) Thanks [@sxzz](https://github.com/sxzz)! - sync version

### Patch Changes

- Updated dependencies [[`0b37cb7`](https://github.com/sxzz/unplugin-vue-macros/commit/0b37cb7e38a5c1893f037fc31f6ca3bd40e74b72)]:
  - @vue-macros/common@0.12.0-beta.2
