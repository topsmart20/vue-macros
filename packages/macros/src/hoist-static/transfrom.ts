import type { TransformContext } from '@vue-macros/common'
import type { Node, Statement } from '@babel/types'

export const transformHoistStatic = (ctx: TransformContext) => {
  function moveToScript(decl: Node, prefix: 'const ' | '' = '') {
    const text = s.slice(startOffset + decl.start!, startOffset + decl.end!)
    s.remove(startOffset + decl.start!, startOffset + decl.end!)
    ctx.scriptCode.prepend += `\n${prefix}${text}`
  }

  const { sfc, s } = ctx
  if (!sfc.scriptSetup) return
  const startOffset = sfc.scriptSetup.loc.start.offset

  for (const stmt of sfc.scriptCompiled!.scriptSetupAst as Statement[]) {
    if (stmt.type === 'VariableDeclaration' && stmt.kind === 'const') {
      const decls = stmt.declarations
      let count = 0
      for (const [i, decl] of decls.entries()) {
        if (!decl.init || !isLiteral(decl.init)) continue

        count++
        moveToScript(decl, 'const ')

        // remove the comma
        if (decls.length > 1) {
          const isLast = i === decls.length - 1
          const start = isLast ? decls[i - 1].end! : decl.end!
          const end = isLast ? decl.start! : decls[i + 1].start!
          s.remove(startOffset + start, startOffset + end)
        }
      }
      if (count === decls.length) {
        s.remove(startOffset + stmt.start!, startOffset + stmt.end!)
      }
    } else if (stmt.type === 'TSEnumDeclaration') {
      const isAllConstant = stmt.members.every(
        (member) => !member.initializer || isLiteral(member.initializer)
      )
      if (!isAllConstant) continue
      moveToScript(stmt)
    }
  }
}

const isLiteral = (node: Node): boolean => {
  // magic comment
  if (
    node.leadingComments?.some(
      (comment) => comment.value.trim() === 'hoist-static'
    )
  ) {
    return true
  }

  switch (node.type) {
    case 'UnaryExpression': // !true
      return isLiteral(node.argument)
    case 'BinaryExpression': // 1 + 2
      return isLiteral(node.left) && isLiteral(node.right)
    case 'ArrayExpression': // [1, 2]
      return node.elements.every((element) => element && isLiteral(element))
    case 'ObjectExpression': // { foo: 1 }
      return node.properties.every(
        (prop) =>
          prop.type === 'ObjectProperty' &&
          (isLiteral(prop.key) ||
            (!prop.computed && prop.key.type === 'Identifier')) &&
          isLiteral(prop.value)
      )
    case 'SequenceExpression': // (1, 2)
    case 'TemplateLiteral': // `123`
      return node.expressions.every((expr) => isLiteral(expr))
  }

  if (isLiteralType(node)) return true
  return false
}

const isLiteralType = (node: Node) => node.type.endsWith('Literal')
