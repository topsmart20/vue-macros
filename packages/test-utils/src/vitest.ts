import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
