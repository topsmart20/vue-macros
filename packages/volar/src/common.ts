import { type Segment, replace } from '@volar/vue-language-core'
import { type FileRangeCapabilities } from '@volar/language-core'

export function getVueLibraryName(vueVersion: number) {
  return vueVersion < 2.7 ? '@vue/runtime-dom' : 'vue'
}

export function addProps(
  content: Segment<FileRangeCapabilities>[],
  decl: Segment<FileRangeCapabilities>[],
  vueLibName: string
) {
  replace(
    content,
    /setup\(\) {/,
    'props: ({} as ',
    ...decl,
    '),\n',
    'setup() {'
  )
  content.push(
    `type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;\n`,
    `type __VLS_TypePropsToRuntimeProps<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? { type: import('${vueLibName}').PropType<__VLS_NonUndefinedable<T[K]>> } : { type: import('${vueLibName}').PropType<T[K]>, required: true } };\n`
  )
  return true
}

export function addEmits(
  content: Segment<FileRangeCapabilities>[],
  decl: Segment<FileRangeCapabilities>[]
) {
  const idx = content.indexOf('setup() {\n')
  if (idx === -1) return false

  replace(
    content,
    /setup\(\) {/,
    'emits: ({} as ',
    ...decl,
    '),\n',
    'setup() {'
  )
  return true
}
