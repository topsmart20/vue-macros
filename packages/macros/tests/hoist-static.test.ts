/// <reference types="vite/client" />

import { describe, expect, test } from 'vitest'
import { finalizeContext, initContext } from '@vue-macros/common'
import { transformHoistStatic } from '../src/hoist-static/transfrom'

describe('define-model', async () => {
  const files = import.meta.glob('./fixtures/hoist-static/*.{vue,js,ts}', {
    eager: true,
    as: 'raw',
  })

  for (const [id, code] of Object.entries(files)) {
    test(id.replace(/\\/g, '/'), async () => {
      const exec = () => {
        const { ctx, getMagicString } = initContext(code, id)
        transformHoistStatic(ctx)
        finalizeContext(ctx)
        return getMagicString()?.toString()
      }
      if (id.includes('error')) {
        expect(exec).toThrowErrorMatchingSnapshot()
      } else {
        expect(exec()).toMatchSnapshot()
      }
    })
  }
})
