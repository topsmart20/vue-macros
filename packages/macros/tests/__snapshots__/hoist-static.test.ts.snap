// Vitest Snapshot v1

exports[`hoist-static > ./fixtures/hoist-static/basic.vue 1`] = `
"<script lang=\\"ts\\">
const a = 'Foo'
const b = 'Foo'!
const c = 1 + 2
const d = 1 || 2
const e = !1
const f = 1 ? 2 : 3
const g = 1 as number
const i = {
  a: 'hello',
  ['b']: 123,
  'd': true,
  nested: {
    a: 'b',
  },
}
</script>
<script setup lang=\\"ts\\">







// const h = <number>2

</script>

<template>
  <div />
</template>
"
`;

exports[`hoist-static > ./fixtures/hoist-static/magic-comment.vue 1`] = `
"<script lang=\\"ts\\">
const name = /* hoist-static */ fn()
</script>
<script setup lang=\\"ts\\">

</script>
"
`;

exports[`hoist-static > ./fixtures/hoist-static/props.vue 1`] = `
"<script lang=\\"ts\\">
const name = 'title'
</script>
<script setup lang=\\"ts\\">


defineProps([name])
</script>
"
`;
